File: AdSwitcher-Android/unityads-adapter/src/main/java/net/adswitcher/adapter/unityads/UnityAdsAdapter.java
package net.adswitcher.adapter.unityads;

import android.app.Activity;
import android.os.Handler;
import android.util.Log;

import com.unity3d.ads.IUnityAdsListener;
import com.unity3d.ads.UnityAds;

import java.util.Map;

import net.adswitcher.adapter.InterstitialAdAdapter;
import net.adswitcher.adapter.InterstitialAdListener;

/**
 * Created by tkyaji on 2016/07/19.
 */
public class UnityAdsAdapter implements InterstitialAdAdapter, IUnityAdsListener {

    private static final String TAG = "UnityAdsAdapter";

    private Activity activity;
    private String placementId;
    private InterstitialAdListener interstitialAdListener;


    @Override
    public void interstitialAdInitialize(Activity activity, InterstitialAdListener interstitialAdListener, Map<String, String> parameters, boolean testMode) {
        this.activity = activity;
        this.interstitialAdListener = interstitialAdListener;

        String gameId = parameters.get("game_id");
        this.placementId = parameters.get("placement_id");
        Log.d(TAG, "videoAdInitialize : game_id=" + gameId + ", placement_id=" + this.placementId);

        UnityAds.initialize(activity, gameId, this, testMode);
        UnityAds.setDebugMode(testMode);
    }

    @Override
    public void interstitialAdLoad() {
        Log.d(TAG, "interstitialAdLoad");

        if (this.isReady()) {
            this.interstitialAdListener.interstitialAdLoaded(UnityAdsAdapter.this, true);

        } else {
            this.adLoad(1);
        }
    }

    private boolean isReady() {
        if (this.placementId == null) {
            return UnityAds.isReady();
        } else {
            return UnityAds.isReady(this.placementId);
        }
    }

    @Override
    public void interstitialAdShow() {
        Log.d(TAG, "interstitialAdShow");

        if (this.placementId == null) {
            UnityAds.show(this.activity);

        } else {
            UnityAds.show(this.activity, this.placementId);
        }
    }


    @Override
    public void onUnityAdsReady(String s) {
        Log.d(TAG, "onUnityAdsReady");
    }

    @Override
    public void onUnityAdsStart(String s) {
        Log.d(TAG, "onUnityAdsStart");
        this.interstitialAdListener.interstitialAdShown(this);
    }

    @Override
    public void onUnityAdsFinish(String s, UnityAds.FinishState finishState) {
        Log.d(TAG, "onUnityAdsFinish : finishState=" + finishState);

        boolean result = (finishState == UnityAds.FinishState.COMPLETED || finishState == UnityAds.FinishState.SKIPPED);
        boolean isSkipped = (finishState == UnityAds.FinishState.SKIPPED);

        this.interstitialAdListener.interstitialAdClosed(this, result, isSkipped);
    }

    @Override
    public void onUnityAdsError(UnityAds.UnityAdsError unityAdsError, String s) {
        Log.d(TAG, "onUnityAdsError : unityAdsError=" + unityAdsError);

    }



    private void adLoad(final int count) {
        Log.d(TAG, "adLoad : count=" + count);

        (new Handler()).postDelayed(new Runnable() {
            @Override
            public void run() {
                if (UnityAdsAdapter.this.isReady()) {
                    UnityAdsAdapter.this.activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            UnityAdsAdapter.this.interstitialAdListener.interstitialAdLoaded(UnityAdsAdapter.this, true);
                        }
                    });
                } else if (count == 5) {
                    UnityAdsAdapter.this.activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            UnityAdsAdapter.this.interstitialAdListener.interstitialAdLoaded(UnityAdsAdapter.this, false);
                        }
                    });
                } else {
                    UnityAdsAdapter.this.adLoad(count + 1);
                }
            }
        }, 1000);
    }

}

File: AdSwitcher-iOS/AdSwitcher/Adapters/UnityAdsAdapter/UnityAdsAdapter.m
//
//  UnityAdsAdapter.m
//  AdSwitcher
//
//  Created by tkyaji on 2016/07/01.
//  Copyright © 2016年 adwitcher. All rights reserved.
//

#import "UnityAdsAdapter.h"

@implementation UnityAdsAdapter {
    UIViewController *_viewController;
    NSString *_placementId;
}

@synthesize interstitialAdDelegate;

#pragma - VideoAdAdapter

- (void)interstitialAdInitialize:(UIViewController *)viewController parameters:(NSDictionary<NSString *,NSString *> *)parameters testMode:(BOOL)testMode {
    NSString *gameId = [parameters objectForKey:@"game_id"];
    _placementId = [parameters objectForKey:@"placement_id"];
    _DLOG("game_id=%@, placement_id=%@", gameId, _placementId);
    
    _viewController = viewController;
    
    [UnityAds initialize:gameId delegate:self testMode:testMode];
}

- (void)interstitialAdLoad {
    if ([self isReady]) {
        _DLOG("ready");
        [self.interstitialAdDelegate interstitialAdLoaded:self result:YES];
        
    } else {
        [self adLoad:1];
    }
}

- (BOOL)isReady {
    if (_placementId) {
        return [UnityAds isReady:_placementId];
    } else {
        return [UnityAds isReady];
    }
}

- (void)interstitialAdShow {
    _DLOG();
    if (_placementId) {
        [UnityAds show:_viewController placementId:_placementId];
    } else {
        [UnityAds show:_viewController];
    }
}



#pragma - UnityAdsDelegate

- (void)unityAdsReady:(NSString *)placementId {
    _DLOG();
}

- (void)unityAdsDidError:(UnityAdsError)error withMessage:(NSString *)message {
    _DLOG("message=%@", message);
}

- (void)unityAdsDidStart:(NSString *)placementId {
    _DLOG();
    [self.interstitialAdDelegate interstitialAdShown:self];
}

- (void)unityAdsDidFinish:(NSString *)placementId withFinishState:(UnityAdsFinishState)state {
    _DLOG("state=%d", (int)state);
    switch (state) {
        case kUnityAdsFinishStateError:
            [self.interstitialAdDelegate interstitialAdClosed:self result:NO isSkipped:NO];
            break;
            
        case kUnityAdsFinishStateCompleted:
            [self.interstitialAdDelegate interstitialAdClosed:self result:YES isSkipped:NO];
            break;
            
        case kUnityAdsFinishStateSkipped:
        [self.interstitialAdDelegate interstitialAdClosed:self result:YES isSkipped:YES];
            break;
    }
}


#pragma - Private methods

- (void)adLoad:(int)count {
    _DLOG("count=%d", count);
    
    dispatch_time_t time = dispatch_time(DISPATCH_TIME_NOW, 1 * NSEC_PER_SEC);
    dispatch_after(time, dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        if ([self isReady]) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.interstitialAdDelegate interstitialAdLoaded:self result:YES];
            });
        } else if (count == 5) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.interstitialAdDelegate interstitialAdLoaded:self result:NO];
            });
        } else {
            [self adLoad:count + 1];
        }
    });
}


@end
